message_ix_models/model/water/build.py

```diff
@@ -3,18 +3,107 @@
 from typing import Mapping
 
 import pandas as pd
+from message_ix import make_df
 from sdmx.model.v21 import Code
 
 from message_ix_models import Context, ScenarioInfo
 from message_ix_models.model import build
 from message_ix_models.model.structure import get_codes
-from message_ix_models.util import package_data_path
+from message_ix_models.util import broadcast, package_data_path
 
 from .utils import read_config
 
 log = logging.getLogger(__name__)
 
 
+def cat_tec_cooling(context: Context) -> tuple[pd.DataFrame, list[str]]:
+    """
+    Categorize cooling technologies based on predefined types and match them with
+    parent technologies present in the scenario.
+
+    This function extracts cooling technology data from a CSV file, filters them
+    based on parent technologies available in the scenario, and categorizes each
+    cooling technology into a predefined type. It also retrieves a list of unique
+    region nodes from the scenario parameter data.
+
+    Parameters
+    ----------
+    context : Context
+        Provides access to the current scenario and configuration.
+
+    Returns
+    -------
+    tuple[pd.DataFrame, list[str]]
+        - cat_tec: A DataFrame with columns:
+            - 'type_tec': Cooling technology category.
+            - 'tec': Name of the cooling technology.
+        - regions_df: A list of unique region nodes from the scenario.
+    """
+    # Define cooling type categories and their corresponding strings
+    cooling_types = {
+        "share_cooling_ot_fresh_tot": ["ot_fresh", "cl_fresh", "air", "ot_saline"],
+        "share_cooling_cl_fresh_tot": ["ot_fresh", "cl_fresh", "air", "ot_saline"],
+        "share_cooling_air_tot": ["ot_fresh", "cl_fresh", "air", "ot_saline"],
+        "share_cooling_ot_saline_tot": ["ot_fresh", "cl_fresh", "air", "ot_saline"],
+        "share_cooling_ot_fresh_share": ["ot_fresh"],
+        "share_cooling_cl_fresh_share": ["cl_fresh"],
+        "share_cooling_air_share": ["air"],
+        "share_cooling_ot_saline_share": ["ot_saline"],
+    }
+
+    FILE = "tech_water_performance_ssp_msg.csv"
+    path = package_data_path("water", "ppl_cooling_tech", FILE)
+    df = pd.read_csv(path)
+    cooling_df = df.loc[df["technology_group"] == "cooling"].copy()
+    # Separate a column for parent technologies of respective cooling
+    # techs
+    cooling_df["parent_tech"] = (
+        cooling_df["technology_name"]
+        .apply(lambda x: pd.Series(str(x).split("__")))
+        .drop(columns=1)
+    )
+    # Extract unique technologies
+    sc = context.get_scenario()
+    #  get df = sc.par("input") for technollgies in cooling_df(parent_tach)
+    df = sc.par("input", filters={"technology": cooling_df["parent_tech"].unique()})
+    missing_tec = cooling_df["parent_tech"][
+        ~cooling_df["parent_tech"].isin(df["technology"])
+    ]
+    # some techs only ahve output, like csp
+    ref_output: pd.DataFrame = sc.par("output", {"technology": missing_tec})
```

Comment:
Does `sc.par()` not return a `pd.DataFrame` already? Why do we need the extra annotation?

Reply:
I am not sure, but also in `water_for_ppl` other parameters were introduced like this, so I just did the same